### PR TITLE
Feature/clojure

### PR DESCRIPTION
**Pull request by matrixjnr**
**What does this PR do?**

Fixed rebuilding Java applications upon second and following commits.
Added support for clojure deployment with `lein` .(To be tested before merging)
Changed the Java sample app from spark to a simple plain Java tcp server with workers. 

**Tasks Completed**

Rebuilding application is done when the `target/build` folder exists

### CODE REVIEW

**Comment by chr15m** [comment]
Wow, nice, keen to check this out!

**Comment by rcarmo** [comment]
Well, someone has to test it... and I barely have connectivity this week ;)

**Comment by chr15m** [comment]
On my todo list.

**Comment by matrixjnr** [comment]
I wish `piku` would have integrated tests to test deployments instead of manual testing. The problem I'm encountering is `lein` dropping into `root` while it already exists in the `piku` home folder.
Basically it's a permission denied creating piku/.lein folder. The `lein` team says that the problem might be the upstart script running as `root`.

**Comment by chr15m** [comment]
@matrixjnr hm weird. Looking at the diff I can't see how it would be run as anything other than the `piku` user. Can you paste console errors?

**Comment by rcarmo** [comment]
> I wish piku would have integrated tests to test deployments instead of manual testing. 
> 
I’m working on a container image to do just that, but it isn’t trivial to set up automated tests for something that effectively does SSH sessions to itself... especially not in volatile cloud-based CI systems...
> The problem I'm encountering is lein dropping into root while it already exists in the piku home folder.
> Basically it's a permission denied creating piku/.lein folder. The lein team says that the problem might be the upstart script running as root.
> 
Nothing in piku runs (or should run) as root. Especially not anything that is a direct result of a push or spawned from the script - if that is happening, then it’s a bug.

I’d lean more towards Leiningen expecting some environment variables to be set, but like Chris said, logs would be great (I can’t test this ATM, e-mail only for a few days)

**Comment by chr15m** [comment]
> I’m working on a container image to do just that, but it isn’t trivial to set up automated tests for something that effectively does SSH sessions to itself... especially not in volatile cloud-based CI systems...

I can imagine this would be complex! Previously I was thinking to test on a local Vagrant virtualbox and I realised it would be impossible to test the Acme SSL cert stuff since it requires a live IP with the correct domain pointing at it.

**Comment by matrixjnr** [comment]
This is what I am getting...
``` bash
jmatrix@jmatrix:~/piku-workspace/sample-clojure-app$ git push piku master
Counting objects: 20, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (14/14), done.
Writing objects: 100% (20/20), 8.27 KiB | 2.07 MiB/s, done.
Total 20 (delta 1), reused 0 (delta 0)
remote: -----> Creating app 'clojuretest'
remote: -----> Deploying app 'clojuretest'
remote: HEAD is now at 7131369 Initial Commit
remote: -----> Checking requirements: ['java', 'lein', 'clojure']
remote: ['/usr/bin/java', '/usr/bin/lein', '/usr/local/bin/clojure']
remote: -----> Clojure app detected.
remote: -----> Building Clojure Application
remote: Downloading Leiningen to /.lein/self-installs/leiningen-2.9.1-standalone.jar now...
remote: mkdir: cannot create directory '/.lein': Permission denied
remote:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
remote:                                  Dload  Upload   Total   Spent    Left  Speed
remote: 100   618    0   618    0     0    295      0 --:--:--  0:00:02 --:--:--   296
remote:   0 13.9M    0     0    0     0      0      0 --:--:--  0:00:04 --:--:--     0Warning: Failed to create the file 
remote: Warning: /.lein/self-installs/leiningen-2.9.1-standalone.jar.pending: No such 
remote: Warning: file or directory
remote:   0 13.9M    0 16360    0     0   3318      0  1:13:26  0:00:04  1:13:22  6805
remote: curl: (23) Failed writing body (0 != 16360)
remote: Failed to download https://github.com/technomancy/leiningen/releases/download/2.9.1/leiningen-2.9.1-standalone.zip (exit code 23)
remote: It's possible your HTTP client's certificate store does not have the
remote: correct certificate authority needed. This is often caused by an
remote: out-of-date version of libssl. It's also possible that you're behind a
remote: firewall and haven't set HTTP_PROXY and HTTPS_PROXY.
remote: -----> nginx NGINX_IPV4_ADDRESS set to 0.0.0.0
remote: -----> nginx NGINX_IPV6_ADDRESS set to [::]
remote: -----> nginx BIND_ADDRESS set to 127.0.0.1
remote: -----> spawning 'clojuretest:jwsgi.1'
remote: -----> spawning 'clojuretest:worker.1'
To jmatrix:clojuretest
 * [new branch]      master -> master
```

The worker is running well since it is directly run by `clojure`.

**Comment by chr15m** [comment]
@matrixjnr Ah ok it seems like Lein is not seeing `$HOME` - maybe try passing it in explicitly to both `lein` invocations by adding it to `env`?

You can do this similarly to here: https://github.com/piku/piku/blob/master/piku.py#L578

Or you could `env.update(environ)` to get all environment variables.

BTW I also noticed the second `lein uberjar` invocation does not have any `env` variable at all passed to it so it won't have any environment context.

**Comment by chr15m** [comment]
@matrixjnr To be more specific, I mean updating `env` with `HOME` here and passing it to both `lein` calls: https://github.com/piku/piku/pull/99/files#diff-bc9247ae52bec05571146b5895110a89R457-R467

**Comment by matrixjnr** [comment]
> @matrixjnr Ah ok it seems like Lein is not seeing `$HOME` - maybe try passing it in explicitly to both `lein` invocations by adding it to `env`?
> 
> You can do this similarly to here: https://github.com/piku/piku/blob/master/piku.py#L578
> 
> Or you could `env.update(environ)` to get all environment variables.
> 
> BTW I also noticed the second `lein uberjar` invocation does not have any `env` variable at all passed to it so it won't have any environment context.

Thanks for the suggestion.

**Comment by matrixjnr** [comment]
@chr15m  I successfully deployed a Clojure Uberjar after setting `LEIN_HOME` as suggested.
Demo can be found here:

[asciicast](https://asciinema.org/a/266379)

I think the PR can be merged.

**Comment by chr15m** [comment]
@matrixjnr this is exciting work, thank you. Can I suggest we break this into 3 different PRs?

 * A PR for the Java example change.
 * A PR for the changes which get Java working.
 * A PR for the changes which get Clojure working.
 * A PR for the [documentation changes like this](https://github.com/piku/piku/pull/99/files#diff-bc9247ae52bec05571146b5895110a89R1010).

I think this will make it easier to merge them as we can test each on in isolation. For example the Java app + documentation changes should be easy to check and can go straight in.

Sorry to cause you more work.

File: piku.py [hunk 1]
```diff
@@ -403,29 +411,64 @@ def deploy_java(app, deltas={}):
...
+    
+    virtual = join(ENV_ROOT, app)
+    target_path = join(APP_ROOT, app, 'target')
+    env_file = join(APP_ROOT, app, 'ENV')
+    projectfile = join(APP_ROOT, app, 'project.clj')
+
+
+    if not exists(target_path):
+        venv = 'mkdir ' + virtual
+        call(venv, cwd=PIKU_ROOT, shell=True)
```
Comment (chr15m) [inline piku.py]: I think it would be good to use `makedirs` here as done here: https://github.com/piku/piku/blob/master/piku.py#L226-L227

File: piku.py [hunk 2]
```diff
@@ -403,29 +411,64 @@ def deploy_java(app, deltas={}):
...
+            "PATH": ':'.join([join(virtual, "bin"), join(app, ".bin"),environ['PATH']]),
+            "LEIN_HOME": environ.get('LEIN_HOME', join(environ['HOME'],'.lein'))
+        }
+        if exists(env_file):
+            env.update(parse_settings(env_file, env))
+            echo("-----> Building Clean Clojure Application")
+            call('lein clean && lein uberjar', cwd=join(APP_ROOT, app), env=env, shell=True)
+        else:
+            echo("-----> Building Clean Clojure Application")
+            call("lein clean && lein uberjar", cwd=join(APP_ROOT, app), env=env, shell=True)
```
Comment (chr15m) [inline piku.py]: This and the call above are the same command - can we collapse these?

File: piku.py [hunk 3]
```diff
@@ -403,29 +411,64 @@ def deploy_java(app, deltas={}):
...
+            echo("-----> Building Clojure Application")
+            call('lein uberjar', cwd=join(APP_ROOT, app), env=env, shell=True)
+        else:
+            echo("-----> Building Clojure Application")
+            call("lein uberjar",cwd=join(APP_ROOT, app), env=env, shell=True)
+    else:
+        env = {
+            'VIRTUAL_ENV': virtual,
+            "PATH": ':'.join([join(virtual, "bin"), join(app, ".bin"),environ['PATH']]),
+            "LEIN_HOME": environ.get('LEIN_HOME', join(environ['HOME'],'.lein'))
```
Comment (chr15m) [inline piku.py]: This env block is duplicated exactly above - can we simplify?

File: piku.py [hunk 4]
```diff
@@ -403,29 +411,64 @@ def deploy_java(app, deltas={}):
...
+        }
+        if exists(env_file):
+            env.update(parse_settings(env_file, env))
+            echo("-----> Building Clojure Application")
+            call('lein uberjar', cwd=join(APP_ROOT, app), env=env, shell=True)
+        else:
+            echo("-----> Building Clojure Application")
+            call("lein uberjar",cwd=join(APP_ROOT, app), env=env, shell=True)
```
Comment (chr15m) [inline piku.py]: This and the call above are the same command - can we collapse these?

File: piku.py [hunk 5]
```diff
@@ -403,29 +411,64 @@ def deploy_java(app, deltas={}):
...
+            'VIRTUAL_ENV': virtual,
+            "PATH": ':'.join([join(virtual, "bin"), join(app, ".bin"),environ['PATH']])
+        }
+        if exists(env_file):
+            env.update(parse_settings(env_file, env))
+            echo("-----> Rebuilding Building Java Application")
+            call('mvn clean package', cwd=join(APP_ROOT, app), env=env, shell=True)
+        else:
+            echo("-----> Rebuilding Java Application")
+            call("mvn clean package", cwd=join(APP_ROOT, app), env=env, shell=True)
```
Comment (chr15m) [inline piku.py]: This and the call above are the same command - can we collapse these?

File: piku.py [hunk 6]
```diff
@@ -373,27 +375,33 @@ def deploy_gradle(app, deltas={}):
...
+        env = {
+            'VIRTUAL_ENV': virtual,
+            "PATH": ':'.join([join(virtual, "bin"), join(app, ".bin"),environ['PATH']])
+        }
+        if exists(env_file):
+            echo ("-----> Rebuilding your Gradle Project")
+            call('gradle clean build', cwd=join(APP_ROOT, app), env=env, shell=True)
+        else:
+            echo("-----> Building Java Application")
+            call('gradle clean build', cwd=join(APP_ROOT, app), env=env, shell=True)
```
Comment (chr15m) [inline piku.py]: This and the call above are the same command - can we collapse these?

File: piku.py [hunk 7]
```diff
@@ -361,8 +364,7 @@ def deploy_gradle(app, deltas={}):
     env_file = join(APP_ROOT, app, 'ENV')
     build = join(APP_ROOT, app, 'build.gradle')
 
-    first_time = False
-    if not exists(target_path) or first_time == True:
+    if not exists(target_path):
         venv = 'mkdir ' + virtual
```
Comment (chr15m) [inline piku.py]: I think it would be good to use makedirs here as done here: https://github.com/piku/piku/blob/master/piku.py#L226-L227

File: piku.py [hunk 8]
```diff
@@ -373,27 +375,33 @@ def deploy_gradle(app, deltas={}):
...
 
     virtual = join(ENV_ROOT, app)
     target_path = join(APP_ROOT, app, 'target')
     env_file = join(APP_ROOT, app, 'ENV')
     pom = join(APP_ROOT, app, 'pom.xml')
 
-    first_time = False
-    if not exists(target_path) or first_time == True:
+    if not exists(target_path):
         venv = 'mkdir ' + virtual
```
Comment (chr15m) [inline piku.py]: I think it would be good to use makedirs here as done here: https://github.com/piku/piku/blob/master/piku.py#L226-L227

**Comment by chr15m** [comment]
Also once the Clojure stuff is merged it would be great to get https://github.com/matrixjnr/sample-clojure-app into the piku org!

File: piku.py (hunk 1 above)
Comment (matrixjnr) [reply piku.py]: Thanks, I will rewrite over the week.

**Comment by rcarmo** [comment]
I just approved the other PR, but am unsure as to what is missing here from @chr15m’s list of split PRs... (not much time lately)

**Comment by chr15m** [comment]
The main stuff from this PR went in in #108 and #104. What's left in this PR is:

 * Some README tweaks mentioning the Java/Clojure stuff.
 * Some doc changes in the `piku.py` runtime which probably need consensus.
 * Removing the `examples/java-spark-app` from this repo.

Note that the Java example currently does not utilise the jwsgi interface correctly. I think it should be redone as a [jwsgi service](https://uwsgi-docs.readthedocs.io/en/latest/JWSGI.html#example) and there should also be a separate vanilla Java HTTP server example behind a simple `web` worker.

*tl;dr*: I think this PR should be closed as any of the above items are probably better done in isolated PRs by @matrixjnr and/or me at some point in the future.

**Comment by den1k** [comment]
The Clojure community is split using both lein and Clojure.tools.deps. Tools.deps is the officially endorsed tool and from what I can see it marks a clear trend in the community. Could support be added for Clojure CLI and tools.deps?

**Comment by matrixjnr** [comment]
@den1k  will check on that. Please create a new issue.
